packages/replayer/src/assets/snapshot-assets.ts

```diff
@@ -14,29 +17,72 @@ const resourceTypesToSnapshot = new Set<ResourceType>([
   assetSnapshots: AssetSnapshot[];
-}) => {
-  return Promise.all(
-    opts.assetSnapshots
-      .filter((snapshot) => snapshot.url.startsWith(opts.baseUrl))
-      .map(async (snapshot) => {
-        const trimmedUrl = withoutQueryParams(snapshot.url).substring(
-          opts.baseUrl.length
-        );
-        const targetFile = join(
-          opts.assetsPath,
-          getFilePath(trimmedUrl, snapshot.contentType)
-        );
-        await mkdir(dirname(targetFile), { recursive: true });
+}
 
-        const data = await snapshot.data;
+const TIMEOUT_FOR_FETCHING_ASSET = Duration.fromObject({ seconds: 5 });
 
-        return writeFile(targetFile, data, { flag: "w" });
-      })
+export const snapshotAssets: (
+  options: SnapshotAssetsOpts
+) => Promise<void> = async (options) => {
+  const { baseUrl, assetSnapshots } = options;
+  await Promise.all(
+    uniqueByRequestUrl(assetSnapshots)
+      .filter((snapshot) => snapshot.url.startsWith(baseUrl))
+      .map((snapshot) => snapshotAssetWithTimeout({ ...options, snapshot }))
   );
+  return;
+};
+
+const snapshotAssetWithTimeout: (
+  options: SnapshotAssetsOpts & { snapshot: AssetSnapshot }
+) => Promise<void> = async (options) => {
+  const { snapshot } = options;
+  const assetPromise = snapshotAsset(options);
+  const timeLimitedPromise = withTimeout(
+    assetPromise,
+    TIMEOUT_FOR_FETCHING_ASSET,
+    `Timed out snapshotting asset for URL ${snapshot.url}`
+  );
+  return timeLimitedPromise.catch((err) => {
+    // We catch and quietly log any errors since asset snapshots
+    // are non-essential/for debugging purposes: we'd rather they fail to
+    // capture, than the CLI fails altogether.
+    log.debug(`Error fetching asset at URL ${snapshot.url}`, err);
+  });
+};
+
+const snapshotAsset: (
+  options: SnapshotAssetsOpts & { snapshot: AssetSnapshot }
+) => Promise<void> = async ({ snapshot, baseUrl, assetsPath }) => {
+  const logger = log.getLogger(METICULOUS_LOGGER_NAME);
+  const trimmedUrl = withoutQueryParams(snapshot.url).substring(baseUrl.length);
+  const targetFile = join(
+    assetsPath,
+    getFilePath(trimmedUrl, snapshot.contentType)
+  );
+  await mkdir(dirname(targetFile), { recursive: true });
+
+  try {
+    const data = await snapshot.getData();
+    return writeFile(targetFile, data, { flag: "w" });
+  } catch (error: unknown) {
+    // There seems to be no other way to tell apart a cancelled request from a normal one
+    // apart from catching the error when trying to read the data (cancelled requests
+    // seem to still have 200 status etc. in Puppeteer oddly). We ignore this, since we don't
+    // care about snapshotting assets for cancelled requests.
+    if ((error as any)?.name === "ProtocolError") {
+      logger.debug(
+        `ProtocolError when fetching snapshotted asset for URL ${snapshot.url}. Ignoring this, ` +
+          `because it's usually due to cancelled requests.`
+      );
+    } else {
+      throw error;
+    }
```

Comment:
nitpick:
```suggestion
    if ((error as any)?.name === "ProtocolError") {
      logger.debug(
        `ProtocolError when fetching snapshotted asset for URL ${snapshot.url}. Ignoring this, ` +
          `because it's usually due to cancelled requests.`
      );
      return;
    }
    
    throw error;
```